astropy/modeling/parameters.py

```diff
@@ -505,8 +505,16 @@ def shape(self):
 
             if model_axis < 0:
                 model_axis = len(shape) + model_axis
-
-            shape = shape[:model_axis] + shape[model_axis + 1:]
+                shape = shape[:model_axis] + shape[model_axis + 1:]
+            else:
+                # When a model set is initialized, the dimension of the parameters
+                # is increased by model_set_axis+1. To find the shape of a parameter
+                # within a single model the extra dimensions need to be removed first.
+                # The following dimension shows the number of models.
+                # The rest of the shape tuple represents the shape of the parameter
+                # in a single model.
+
+                shape = shape[model_axis + 1:]
 
```

Comment:
@perrygreenfield I changed a bit the computation of `Parameter.shape`. Assuming that if model_set_axis is negative then the parameter is an array - went back  to the original computation. Is this a reasonable assumption?
That allows tests with negative axis to pass.

Reply:
Yes, it looks good to me! I'm ok with merging now. Any other issues I'm not aware of?

Reply:
No, that's all. Thanks for the review!
I rebased and tests should pass now.
Could you just approve it formally to allow merging it.

Reply:
Sorry if I'm just adding noise to the conversation, but why should `model_set_axis` being positive or negative depend on whether parameters are arrays?

Reply:
I imagine it's most useful to have a negative axis value if the parameter is an array but let me know if you disagree. From my point of view the code has inconsistencies and ambiguities. The purpose of this PR is to patch the current problems with use cases which are common without changing behavior or introducing features which we know now won't be supported in the future. Fixing it completely will require more extensive changes and more time (though I think they should be done for this coming release). More specifically I think there may be a need for an intrinsic parameter shape which is not computed but set at initialization (perhaps from the default value). I think this is an improvement of the current situation and I don't want to hold other work.

In any case if you think this should not be merged now is the time to speak.

Reply:
Oh, sorry, I think I misunderstood what you were saying.

Reply:
Our comments crossed, but that sounds sensible. 

I was also wondering about setting something at initialization, given the apparent ambiguity (though I should probably stop commenting on things like parameter shapes without studying them properly, now you have a more qualified reviewer...).

Thanks, @perrygreenfield for wading in and @nden for the effort on this!